File: c-claw/src/gateway.c
#include "../include/claw.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <signal.h>

static volatile int g_running = 0;

static void signal_handler(int sig) {
    (void)sig;
    g_running = 0;
}

static char *read_request(int client_fd) {
    char *buffer = malloc(8192);
    if (!buffer) return NULL;

    ssize_t total = 0;
    ssize_t n;

    while (total < 8191) {
        n = recv(client_fd, buffer + total, 8191 - total, 0);
        if (n <= 0) break;
        total += n;

        if (total >= 4 && strstr(buffer, "\r\n\r\n")) {
            const char *cl = strstr(buffer, "Content-Length:");
            if (cl) {
                int content_len = atoi(cl + 15);
                const char *body = strstr(buffer, "\r\n\r\n");
                if (body) {
                    body += 4;
                    int body_received = total - (body - buffer);
                    while (body_received < content_len) {
                        n = recv(client_fd, buffer + total, 8191 - total, 0);
                        if (n <= 0) break;
                        total += n;
                        body_received += n;
                    }
                }
            }
            break;
        }
    }

    buffer[total] = 0;
    return buffer;
}

static void send_response(int client_fd, const char *status, const char *content_type, const char *body) {
    char header[512];
    int body_len = body ? strlen(body) : 0;

    snprintf(header, sizeof(header),
             "HTTP/1.1 %s\r\n"
             "Content-Type: %s\r\n"
             "Content-Length: %d\r\n"
             "Connection: close\r\n"
             "\r\n",
             status, content_type, body_len);

    send(client_fd, header, strlen(header), 0);
    if (body) {
        send(client_fd, body, body_len, 0);
    }
}

static void handle_health(int client_fd) {
    const char *response = "{\"status\":\"ok\"}";
    send_response(client_fd, "200 OK", "application/json", response);
}

static void handle_chat(int client_fd, const char *request) {
    const char *body = strstr(request, "\r\n\r\n");
    if (!body) {
        send_response(client_fd, "400 Bad Request", "text/plain", "Missing request body");
        return;
    }
    body += 4;

    char *prompt = str_extract_field(body, "prompt");
    if (!prompt) {
        send_response(client_fd, "400 Bad Request", "text/plain", "Missing prompt field");
        return;
    }

    Config cfg;
    if (config_load(&cfg) != 0) {
        free(prompt);
        send_response(client_fd, "500 Internal Server Error", "text/plain", "Configuration error");
        return;
    }

    Message msg;
    strcpy(msg.role, "user");
    msg.content = prompt;

    char *reply = provider_chat(&msg, 1);
    free(prompt);

    if (!reply) {
        send_response(client_fd, "500 Internal Server Error", "text/plain", "Provider error");
        return;
    }

    char *escaped = str_escape_json(reply);
    free(reply);

    if (!escaped) {
        send_response(client_fd, "500 Internal Server Error", "text/plain", "Memory error");
        return;
    }

    char response[4096];
    snprintf(response, sizeof(response), "{\"response\":\"%s\"}", escaped);
    free(escaped);

    send_response(client_fd, "200 OK", "application/json", response);
}

static void handle_client(int client_fd) {
    char *request = read_request(client_fd);
    if (!request) {
        close(client_fd);
        return;
    }

    if (strncmp(request, "GET /health", 11) == 0) {
        handle_health(client_fd);
    } else if (strncmp(request, "POST /chat", 10) == 0) {
        handle_chat(client_fd, request);
    } else {
        send_response(client_fd, "404 Not Found", "text/plain", "Not Found");
    }

    free(request);
    close(client_fd);
}

int gateway_start(int port) {
    int server_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (server_fd < 0) {
        perror("socket");
        return -1;
    }

    int opt = 1;
    if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)) < 0) {
        perror("setsockopt");
        close(server_fd);
        return -1;
    }

    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = INADDR_ANY;
    addr.sin_port = htons(port);

    if (bind(server_fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        perror("bind");
        close(server_fd);
        return -1;
    }

    if (listen(server_fd, 10) < 0) {
        perror("listen");
        close(server_fd);
        return -1;
    }

    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);
    g_running = 1;

    printf("Gateway listening on port %d\n", port);

    while (g_running) {
        struct sockaddr_in client_addr;
        socklen_t client_len = sizeof(client_addr);

        int client_fd = accept(server_fd, (struct sockaddr *)&client_addr, &client_len);
        if (client_fd < 0) {
            if (g_running) perror("accept");
            continue;
        }

        handle_client(client_fd);
    }

    close(server_fd);
    return 0;
}

void gateway_stop(void) {
    g_running = 0;
}

File: py-claw/src/gateway.py
"""
gateway.py - 基于标准库的极简 HTTP 服务器
只支持两个端点：GET /health  POST /chat
"""
import sys
import socket
import threading
from typing import Optional


# ── 请求分包接收 ──────────────────────────────────────────────

def _recv_request(conn: socket.socket) -> str:
    """循环读取直到获得完整 HTTP 请求"""
    buf = b""
    conn.settimeout(5.0)
    try:
        while True:
            chunk = conn.recv(4096)
            if not chunk:
                break
            buf += chunk
            # headers 结束标志
            if b"\r\n\r\n" not in buf:
                continue
            header_end = buf.index(b"\r\n\r\n") + 4
            headers_raw = buf[:header_end].decode("utf-8", errors="replace")
            # 查找 Content-Length
            body_received = buf[header_end:]
            content_length = 0
            for line in headers_raw.splitlines():
                if line.lower().startswith("content-length:"):
                    try:
                        content_length = int(line.split(":", 1)[1].strip())
                    except ValueError:
                        pass
            if len(body_received) >= content_length:
                break
    except (socket.timeout, OSError):
        pass
    return buf.decode("utf-8", errors="replace")


# ── 响应构建 ──────────────────────────────────────────────────

def _response(status: str, body: str, content_type: str = "application/json") -> bytes:
    encoded = body.encode("utf-8")
    header = (
        f"HTTP/1.1 {status}\r\n"
        f"Content-Type: {content_type}; charset=utf-8\r\n"
        f"Content-Length: {len(encoded)}\r\n"
        f"Connection: close\r\n"
        "\r\n"
    )
    return header.encode("utf-8") + encoded


# ── 极简路由 ──────────────────────────────────────────────────

def _handle(raw: str, provider_cfg) -> bytes:
    """直接字符串比对做路由，不用路由表"""
    first_line = raw.split("\r\n", 1)[0] if raw else ""

    # GET /health
    if first_line.startswith("GET /health"):
        return _response("200 OK", '{"status":"ok"}')

    # POST /chat
    if first_line.startswith("POST /chat"):
        # 提取 body（headers 之后的部分）
        body = ""
        if "\r\n\r\n" in raw:
            body = raw.split("\r\n\r\n", 1)[1]

        # 从 body 中提取 prompt 字段（手写，只找 "prompt":"..."）
        prompt = _extract_prompt(body)
        if not prompt:
            return _response("400 Bad Request", '{"error":"missing prompt"}')

        from .claw import Message
        from .provider import chat
        response = chat([Message(role="user", content=prompt)], provider_cfg)
        if response is None:
            return _response("500 Internal Server Error", '{"error":"provider error"}')

        # 手写 JSON 转义响应
        from .provider import _escape_json
        return _response("200 OK", f'{{"response":"{_escape_json(response)}"}}')

    return _response("404 Not Found", '{"error":"not found"}')


def _extract_prompt(body: str) -> Optional[str]:
    """从请求 body 里提取 prompt 字符串值"""
    key = '"prompt"'
    pos = body.find(key)
    if pos == -1:
        return None
    pos += len(key)
    while pos < len(body) and body[pos] in (' ', '\t', '\n', '\r', ':'):
        pos += 1
    if pos >= len(body) or body[pos] != '"':
        return None
    pos += 1
    out = []
    while pos < len(body):
        ch = body[pos]
        if ch == '\\':
            pos += 1
            if pos < len(body):
                esc = body[pos]
                if esc == 'n':
                    out.append('\n')
                elif esc == 'r':
                    out.append('\r')
                elif esc == 't':
                    out.append('\t')
                elif esc == '\\':
                    out.append('\\')
                elif esc == '"':
                    out.append('"')
                else:
                    out.append(esc)
        elif ch == '"':
            break
        else:
            out.append(ch)
        pos += 1
    return "".join(out) or None


# ── 服务器主循环 ──────────────────────────────────────────────

_stop_flag = threading.Event()


def start(port: int, provider_cfg) -> int:
    """启动 HTTP 服务，阻塞直到 Ctrl+C"""
    global _stop_flag
    _stop_flag.clear()

    srv = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    srv.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    try:
        srv.bind(("0.0.0.0", port))  # 绑定所有网络接口（网关服务器的预期行为）
    except OSError as e:
        print(f"Error: 无法绑定端口 {port} - {e}", file=sys.stderr)
        return 1
    srv.listen(16)
    srv.settimeout(1.0)

    try:
        while not _stop_flag.is_set():
            try:
                conn, _ = srv.accept()
            except socket.timeout:
                continue
            except OSError:
                break
            t = threading.Thread(target=_serve_conn, args=(conn, provider_cfg), daemon=True)
            t.start()
    except KeyboardInterrupt:
        pass
    finally:
        srv.close()
    return 0


def stop() -> None:
    """发出停止信号"""
    _stop_flag.set()


def _serve_conn(conn: socket.socket, provider_cfg) -> None:
    """在独立线程中处理单个连接"""
    try:
        raw = _recv_request(conn)
        resp = _handle(raw, provider_cfg)
        conn.sendall(resp)
    except OSError:
        pass
    finally:
        try:
            conn.close()
        except OSError:
            pass
